File: sdks/python/sdk/src/moss/client/moss_client.py
from __future__ import annotations

import asyncio
import logging
import os
import threading
import uuid
from collections.abc import Sequence
from typing import Any, ClassVar

import httpx
from moss_core import (
    CLOUD_API_MANAGE_URL,
    DocumentInfo,
    GetDocumentsOptions,
    IndexInfo,
    IndexManager,
    JobStatusResponse,
    ManageClient,
    MutationOptions,
    MutationResult,
    QueryResultDocumentInfo,
    SearchResult,
)

logger = logging.getLogger(__name__)


class QueryOptions:
    """Options for search queries."""

    def __init__(
        self,
        embedding: Sequence[float] | None = None,
        top_k: int | None = None,
        alpha: float | None = None,
        filter: dict | None = None,
        rerank: bool = False,
        rerank_top_k: int | None = None,
        rerank_model: str | None = None,
    ):
        if top_k is not None and (not isinstance(top_k, int) or top_k < 1):
            raise ValueError("top_k must be an integer >= 1")
        if alpha is not None and (
            not isinstance(alpha, (int, float)) or not (0.0 <= alpha <= 1.0)
        ):
            raise ValueError("alpha must be a float between 0.0 and 1.0")
        if embedding is not None:
            try:
                embedding = [float(x) for x in embedding]
            except (TypeError, ValueError):
                raise ValueError("embedding must be a sequence of numbers")
        if rerank_top_k is not None and (
            not isinstance(rerank_top_k, int) or rerank_top_k < 1
        ):
            raise ValueError("rerank_top_k must be an integer >= 1")
        if top_k is not None and rerank_top_k is not None and rerank_top_k < top_k:
            raise ValueError("rerank_top_k must be >= top_k")

        self.embedding = embedding
        self.top_k = top_k
        self.alpha = float(alpha) if alpha is not None else None
        self.filter = filter
        self.rerank = bool(rerank)
        self.rerank_top_k = rerank_top_k
        self.rerank_model = rerank_model


def _get_manage_url() -> str:
    """Manage URL, overridable via env for local development."""
    return os.getenv("MOSS_CLOUD_API_MANAGE_URL", CLOUD_API_MANAGE_URL)


def _get_query_url() -> str:
    """Query URL, derived from manage URL or overridable via env."""
    explicit = os.getenv("MOSS_CLOUD_QUERY_URL")
    if explicit:
        return explicit
    return _get_manage_url().replace("/v1/manage", "/query")


class MossClient:
    """
    Semantic search client for vector similarity operations.

    All mutations and reads go through the Rust ManageClient.
    Querying uses the local IndexManager when the index is loaded,
    otherwise falls back to the cloud query API.

    Example:
        ```python
        from moss import MossClient, DocumentInfo

        client = MossClient("project-id", "project-key")

        docs = [DocumentInfo(id="1", text="Machine learning fundamentals")]
        result = await client.create_index("my-index", docs, "moss-minilm")

        await client.load_index("my-index")
        results = await client.query("my-index", "AI and neural networks")
        ```
    """

    DEFAULT_MODEL_ID = "moss-minilm"
    _cross_encoder_cache: ClassVar[dict[str, Any]] = {}
    _cross_encoder_locks: ClassVar[dict[str, threading.Lock]] = {}
    _cache_lock: ClassVar[threading.Lock] = threading.Lock()

    def __init__(self, project_id: str, project_key: str) -> None:
        self._project_id = project_id
        self._project_key = project_key
        self._client_id = str(uuid.uuid4())
        manage_url = _get_manage_url()
        self._manage = ManageClient(
            project_id, project_key, manage_url, self._client_id
        )
        self._manager = IndexManager(
            project_id, project_key, manage_url, self._client_id
        )

    # -- Mutations (via Rust ManageClient) --------------------------

    async def create_index(
        self,
        name: str,
        docs: list[DocumentInfo],
        model_id: str | None = None,
    ) -> MutationResult:
        """Create a new index and populate it with documents."""
        resolved_model_id = self._resolve_model_id(docs, model_id)
        return await asyncio.to_thread(
            self._manage.create_index,
            name,
            docs,
            resolved_model_id,
        )

    async def add_docs(
        self,
        name: str,
        docs: list[DocumentInfo],
        options: MutationOptions | None = None,
    ) -> MutationResult:
        """Add or update documents in an index."""
        return await asyncio.to_thread(
            self._manage.add_docs,
            name,
            docs,
            options,
        )

    async def delete_docs(
        self,
        name: str,
        doc_ids: list[str],
    ) -> MutationResult:
        """Delete documents from an index by their IDs."""
        return await asyncio.to_thread(
            self._manage.delete_docs,
            name,
            doc_ids,
        )

    async def get_job_status(self, job_id: str) -> JobStatusResponse:
        """Get the status of a bulk operation job."""
        return await asyncio.to_thread(self._manage.get_job_status, job_id)

    # -- Read operations (via Rust ManageClient) --------------------

    async def get_index(self, name: str) -> IndexInfo:
        """Get information about a specific index."""
        return await asyncio.to_thread(self._manage.get_index, name)

    async def list_indexes(self) -> list[IndexInfo]:
        """List all indexes with their information."""
        return await asyncio.to_thread(self._manage.list_indexes)

    async def delete_index(self, name: str) -> bool:
        """Delete an index and all its data."""
        return await asyncio.to_thread(self._manage.delete_index, name)

    async def get_docs(
        self,
        name: str,
        options: GetDocumentsOptions | None = None,
    ) -> list[DocumentInfo]:
        """Retrieve documents from an index."""
        return await asyncio.to_thread(self._manage.get_docs, name, options)

    # -- Index loading & querying -----------------------------------

    async def load_index(
        self,
        name: str,
        auto_refresh: bool = False,
        polling_interval_in_seconds: int = 600,
    ) -> str:
        """
        Downloads an index from the cloud into memory for fast local querying.

        Without load_index(), query() falls back to the cloud API (~100-500ms).
        With load_index(), queries run entirely in-memory (~1-10ms).
        """
        try:
            await asyncio.to_thread(
                self._manager.load_index,
                name,
                auto_refresh,
                polling_interval_in_seconds,
            )
            await asyncio.to_thread(self._manager.load_query_model, name)
            return name
        except RuntimeError as e:
            raise RuntimeError(f"Failed to load index '{name}': {e}") from e

    async def unload_index(self, name: str) -> None:
        """Unload an index from memory."""
        try:
            await asyncio.to_thread(self._manager.unload_index, name)
        except RuntimeError as e:
            raise RuntimeError(f"Failed to unload index '{name}': {e}") from e

    async def query(
        self,
        name: str,
        query: str,
        options: QueryOptions | None = None,
    ) -> SearchResult:
        """
        Perform a semantic similarity search.

        If the index is loaded locally (via load_index), queries run in-memory.
        Otherwise, falls back to the cloud query API.

        Args:
            options: Query options (top_k, alpha, embedding, filter, rerank, etc).
        """
        is_loaded = await asyncio.to_thread(self._manager.has_index, name)

        rerank = getattr(options, "rerank", False) is True
        override_top_k = None
        if rerank:
            candidate_pool = getattr(options, "rerank_top_k", None) or 50
            final_top_k = getattr(options, "top_k", None) or 5
            override_top_k = max(candidate_pool, final_top_k)

        if is_loaded:
            result = await self._query_local(name, query, options, override_top_k)
        else:
            if getattr(options, "filter", None) is not None:
                logger.warning(
                    "Metadata filter ignored: filtering is only supported for locally loaded indexes. "
                    "Call load_index('%s') first.",
                    name,
                )
            result = await self._query_cloud(name, query, options, override_top_k)

        if rerank:
            result = await self._rerank_results(query, result, options)

        return result

    # -- Internal ---------------------------------------------------

    async def _query_local(
        self,
        name: str,
        query: str,
        options: QueryOptions | None,
        override_top_k: int | None = None,
    ) -> SearchResult:
        top_k = (
            override_top_k
            if override_top_k is not None
            else (getattr(options, "top_k", None) or 5)
        )
        alpha = getattr(options, "alpha", None)
        if alpha is None:
            alpha = 0.8
        query_embedding = getattr(options, "embedding", None)
        filter = getattr(options, "filter", None)

        if query_embedding is None:
            try:
                return await asyncio.to_thread(
                    self._manager.query_text,
                    name,
                    query,
                    top_k,
                    alpha,
                    filter,
                )
            except RuntimeError as e:
                if "requires explicit query embeddings" in str(e):
                    raise ValueError(
                        "This index uses custom embeddings. "
                        "Query embeddings must be provided via QueryOptions.embedding."
                    ) from e
                raise

        return await asyncio.to_thread(
            self._manager.query,
            name,
            query,
            list(query_embedding),
            top_k,
            alpha,
            filter,
        )

    async def _query_cloud(
        self,
        name: str,
        query: str,
        options: QueryOptions | None,
        override_top_k: int | None = None,
    ) -> SearchResult:
        """Fallback: query via the cloud API when the index is not loaded locally."""
        top_k = (
            override_top_k
            if override_top_k is not None
            else (getattr(options, "top_k", None) or 10)
        )
        query_embedding = getattr(options, "embedding", None)

        request_body: dict[str, Any] = {
            "query": query,
            "indexName": name,
            "projectId": self._project_id,
            "projectKey": self._project_key,
            "topK": top_k,
        }
        if query_embedding is not None:
            request_body["queryEmbedding"] = list(query_embedding)

        try:
            async with httpx.AsyncClient(timeout=60.0) as client:
                response = await client.post(
                    _get_query_url(),
                    headers={"Content-Type": "application/json"},
                    json=request_body,
                )
                if not response.is_success:
                    raise Exception(f"HTTP error! status: {response.status_code}")
                data = response.json()
        except httpx.RequestError as error:
            raise Exception(f"Cloud query request failed: {error!s}")

        return self._dict_to_search_result(data)

    @staticmethod
    def _dict_to_search_result(data: dict) -> SearchResult:
        docs = [
            QueryResultDocumentInfo(
                id=d.get("id", ""),
                text=d.get("text", ""),
                metadata=d.get("metadata"),
                score=float(d.get("score", 0.0)),
            )
            for d in data.get("docs", [])
        ]
        return SearchResult(
            docs=docs,
            query=data.get("query", ""),
            index_name=data.get("indexName"),
            time_taken_ms=data.get("timeTakenMs"),
        )

    async def _rerank_results(
        self, query: str, search_result: SearchResult, options: QueryOptions | None
    ) -> SearchResult:
        if not search_result.docs:
            return search_result

        try:
            from sentence_transformers import CrossEncoder
        except ImportError:
            raise ImportError(
                "The 'sentence-transformers' package is required for reranking. "
                "Install it with: pip install 'moss[rerank]'"
            )

        model_name = (
            getattr(options, "rerank_model", None)
            or "cross-encoder/ms-marco-MiniLM-L-6-v2"
        )

        def do_rerank() -> SearchResult:
            with self.__class__._cache_lock:
                if model_name not in self.__class__._cross_encoder_locks:
                    self.__class__._cross_encoder_locks[model_name] = threading.Lock()

            model_lock = self.__class__._cross_encoder_locks[model_name]

            with model_lock:
                if model_name not in self.__class__._cross_encoder_cache:
                    self.__class__._cross_encoder_cache[model_name] = CrossEncoder(
                        model_name
                    )

                model = self.__class__._cross_encoder_cache[model_name]

                local_docs = search_result.docs
                pairs = [[query, doc.text] for doc in local_docs]
                scores = model.predict(pairs)

            for doc, score in zip(local_docs, scores):
                doc.score = float(score)

            local_docs.sort(key=lambda d: d.score, reverse=True)

            original_top_k = getattr(options, "top_k", None) or 5
            search_result.docs = local_docs[:original_top_k]
            return search_result

        return await asyncio.to_thread(do_rerank)

    def _resolve_model_id(
        self,
        docs: list[DocumentInfo],
        model_id: str | None,
    ) -> str:
        if model_id is not None:
            return model_id
        has_embeddings = any(
            getattr(doc, "embedding", None) is not None for doc in docs
        )
        return "custom" if has_embeddings else self.DEFAULT_MODEL_ID

File: sdks/javascript/sdk/src/client/mossClient.ts
import { InternalMossClient } from "./internalMossClient";
import type {
  SearchResult,
  IndexInfo,
  DocumentInfo,
  GetDocumentsOptions,
  QueryOptions,
  MutationResult,
  JobStatusResponse,
} from "@moss-dev/moss-core";
import {
  LoadIndexOptions,
  CreateIndexOptions,
  MutationOptions,
  MossModel,
} from "../models";

const DEFAULT_MODEL_ID: MossModel = "moss-minilm";

/**
 * MossClient - Async-first semantic search client for vector similarity operations.
 *
 * All mutations (createIndex, addDocs, deleteDocs) are async operations
 * that run server-side and poll until complete.
 *
 * @example
 * ```typescript
 * import { MossClient } from '@moss-dev/moss';
 *
 * const client = new MossClient('your-project-id', 'your-project-key');
 *
 * // Create an index with documents (polls until complete)
 * const result = await client.createIndex('docs', [
 *   { id: '1', text: 'Machine learning fundamentals' },
 *   { id: '2', text: 'Deep learning neural networks' }
 * ]);
 *
 * // Add docs (polls until complete)
 * await client.addDocs('docs', [
 *   { id: '3', text: 'Natural language processing' }
 * ]);
 *
 * // Query the index
 * await client.loadIndex('docs');
 * const results = await client.query('docs', 'AI and neural networks');
 * ```
 */
export class MossClient {
  #internal: InternalMossClient;

  /**
   * Creates a new MossClient instance.
   * @param projectId - Your project identifier.
   * @param projectKey - Your project authentication key.
   */
  constructor(projectId: string, projectKey: string) {
    this.#internal = new InternalMossClient(projectId, projectKey);
  }


  /**
   * Creates a new index with the provided documents via async upload.
   *
   * Handles the full flow: init → upload → build → poll until complete.
   * Returns when the index is ready.
   *
   * When all documents have pre-computed embeddings, they are serialized as raw
   * float32 in the binary upload. When no documents have embeddings, the server
   * generates embeddings in batches (dimension=0 flow).
   *
   * Mixed documents (some with embeddings, some without) are rejected.
   *
   * @param indexName - Name of the index to create.
   * @param docs - Documents, optionally with pre-computed embeddings.
   * @param options - Optional model ID and progress callback.
   * @returns Promise that resolves to MutationResult when the index is ready.
   * @throws {Error} If the index already exists or creation fails.
   *
   * @example
   * ```typescript
   * const result = await client.createIndex('knowledge-base', [
   *   { id: 'doc1', text: 'Introduction to AI' },
   *   { id: 'doc2', text: 'Machine learning basics' }
   * ], {
   *   onProgress: (p) => console.log(`${p.status} ${p.progress}%`),
   * });
   * ```
   */
  async createIndex(
    indexName: string,
    docs: DocumentInfo[],
    options?: CreateIndexOptions,
  ): Promise<MutationResult> {
    let resolvedModelId: string;
    if (options?.modelId !== undefined) {
      resolvedModelId = options.modelId;
    } else {
      const hasEmbeddings = docs.some(
        (doc) => Array.isArray(doc.embedding) && doc.embedding.length > 0,
      );
      resolvedModelId = hasEmbeddings ? "custom" : DEFAULT_MODEL_ID;
    }
    return this.#internal.createIndex(indexName, docs, resolvedModelId, options);
  }

  /**
   * Gets information about a specific index.
   *
   * @param indexName - Name of the index to retrieve.
   * @returns Promise that resolves to IndexInfo object.
   * @throws {Error} If the index does not exist.
   *
   * @example
   * ```typescript
   * const info = await client.getIndex('knowledge-base');
   * console.log(`Index has ${info.docCount} documents`);
   * ```
   */
  async getIndex(indexName: string): Promise<IndexInfo> {
    return this.#internal.getIndex(indexName);
  }

  /**
   * Lists all available indexes.
   *
   * @returns Promise that resolves to array of IndexInfo objects.
   *
   * @example
   * ```typescript
   * const indexes = await client.listIndexes();
   * indexes.forEach(index => {
   *   console.log(`${index.name}: ${index.docCount} docs`);
   * });
   * ```
   */
  async listIndexes(): Promise<IndexInfo[]> {
    return this.#internal.listIndexes();
  }

  /**
   * Deletes an index and all its data.
   *
   * @param indexName - Name of the index to delete.
   * @returns Promise that resolves to true if successful.
   * @throws {Error} If the index does not exist.
   *
   * @example
   * ```typescript
   * const deleted = await client.deleteIndex('old-index');
   * ```
   */
  async deleteIndex(indexName: string): Promise<boolean> {
    return this.#internal.deleteIndex(indexName);
  }


  /**
   * Adds or updates documents in an index asynchronously.
   *
   * The index rebuild happens server-side. This method polls until
   * the rebuild is complete and then returns.
   *
   * @param indexName - Name of the target index.
   * @param docs - Documents to add or update.
   * @param options - Optional configuration (upsert, onProgress callback).
   * @returns Promise that resolves to MutationResult when the operation is complete.
   * @throws {Error} If the index does not exist.
   *
   * @example
   * ```typescript
   * const result = await client.addDocs('knowledge-base', [
   *   { id: 'new-doc', text: 'New content to index' }
   * ], { upsert: true });
   * console.log(`Job ${result.jobId} completed`);
   * ```
   */
  async addDocs(
    indexName: string,
    docs: DocumentInfo[],
    options?: MutationOptions,
  ): Promise<MutationResult> {
    return this.#internal.addDocs(indexName, docs, options);
  }

  /**
   * Deletes documents from an index by their IDs asynchronously.
   *
   * The index rebuild happens server-side. This method polls until
   * the rebuild is complete and then returns.
   *
   * @param indexName - Name of the target index.
   * @param docIds - Array of document IDs to delete.
   * @param options - Optional configuration (onProgress callback).
   * @returns Promise that resolves to MutationResult when the operation is complete.
   * @throws {Error} If the index does not exist.
   *
   * @example
   * ```typescript
   * const result = await client.deleteDocs('knowledge-base', ['doc1', 'doc2']);
   * console.log(`Job ${result.jobId} completed`);
   * ```
   */
  async deleteDocs(
    indexName: string,
    docIds: string[],
    options?: MutationOptions,
  ): Promise<MutationResult> {
    return this.#internal.deleteDocs(indexName, docIds, options);
  }


  /**
   * Gets the current status of an async job.
   *
   * @param jobId - The job ID returned by createIndex, addDocs, or deleteDocs.
   * @returns Promise that resolves to JobStatusResponse with progress details.
   *
   * @example
   * ```typescript
   * const status = await client.getJobStatus(jobId);
   * console.log(`${status.status} — ${status.progress}%`);
   * ```
   */
  async getJobStatus(jobId: string): Promise<JobStatusResponse> {
    return this.#internal.getJobStatus(jobId);
  }


  /**
   * Retrieves documents from an index.
   *
   * @param indexName - Name of the target index.
   * @param options - Optional configuration for retrieval.
   * @returns Promise that resolves to array of documents.
   * @throws {Error} If the index does not exist.
   *
   * @example
   * ```typescript
   * // Get all documents
   * const allDocs = await client.getDocs('knowledge-base');
   *
   * // Get specific documents
   * const specificDocs = await client.getDocs('knowledge-base', {
   *   docIds: ['doc1', 'doc2']
   * });
   * ```
   */
  async getDocs(
    indexName: string,
    options?: GetDocumentsOptions,
  ): Promise<DocumentInfo[]> {
    return this.#internal.getDocs(indexName, options);
  }


  /**
   * Downloads an index from the cloud into memory for fast local querying.
   *
   * **How it works:**
   * 1. Fetches the index assets from the cloud
   * 2. Loads the embedding model for generating query embeddings
   * 3. Executes a local similarity match between the query embedding and the retrieved index.
   *
   * **Why use this?**
   * - Without `loadIndex()`: Every `query()` call goes to the cloud API (~100-500ms network latency)
   * - With `loadIndex()`: Queries run entirely in-memory (~1-10ms)
   *
   * **Reload behavior:**
   * If the index is already loaded, calling `loadIndex()` again will:
   * - Stop any existing auto-refresh polling
   * - Download a fresh copy from the cloud
   * - Replace the in-memory index
   *
   * **Auto-refresh (optional):**
   * Enable `autoRefresh: true` to periodically poll the cloud for updates.
   * When a newer version is detected, the index is automatically hot-swapped
   * without interrupting queries.
   *
   * @param indexName - Name of the index to load.
   * @param options - Optional configuration including auto-refresh settings.
   * @returns Promise that resolves to the index name.
   * @throws {Error} If the index does not exist in the cloud or loading fails.
   *
   * @example
   * ```typescript
   * // Simple load - enables fast local queries
   * await client.loadIndex('my-index');
   *
   * // Now queries run locally (fast, no network calls)
   * const results = await client.query('my-index', 'search text');
   *
   * // Load with auto-refresh to keep index up-to-date
   * await client.loadIndex('my-index', {
   *   autoRefresh: true,
   *   pollingIntervalInSeconds: 300, // Check cloud every 5 minutes
   * });
   *
   * // Stop auto-refresh by reloading without the option
   * await client.loadIndex('my-index');
   * ```
   */
  async loadIndex(indexName: string, options?: LoadIndexOptions): Promise<string> {
    return this.#internal.loadIndex(indexName, options);
  }

  /**
   * Performs a semantic similarity search against the specified index.
   *
   * If the index has been loaded via `loadIndex()`, runs entirely in-memory.
   * Otherwise, falls back to the cloud `/query` endpoint.
   *
   * @param indexName - Name of the target index to search.
   * @param query - The search query text.
   * @param options - Optional query configuration including topK (default: 5) and embedding overrides.
   * @returns Promise that resolves to SearchResult with matching documents.
   * @throws {Error} If the specified index does not exist.
   *
   * @example
   * ```typescript
   * const results = await client.query('knowledge-base', 'machine learning');
   * results.docs.forEach(doc => {
   *   console.log(`${doc.id}: ${doc.text} (score: ${doc.score})`);
   * });
   * ```
   */
  async query(
    indexName: string,
    query: string,
    options?: QueryOptions,
  ): Promise<SearchResult> {
    return this.#internal.query(indexName, query, options);
  }

}
